tests/test_other.py

```diff
@@ -7596,6 +7596,57 @@ def test_binaryen_ignore_implicit_traps(self):
       print('sizes:', sizes)
       assert sizes[1] < sizes[0], 'ignoring implicit traps must reduce code size'
 
+  def test_binaryen_methods(self):
+    for method_init in ['interpret-asm2wasm', 'interpret-s-expr', 'asmjs', 'interpret-binary', 'asmjs,interpret-binary', 'interpret-binary,asmjs']:
+      # check success and failure for simple modes, only success for combined/fallback ones
+      for success in [1, 0] if ',' not in method_init else [1]:
+        method = method_init
+        command = [PYTHON, EMCC, '-o', 'a.wasm.js', '-s', 'BINARYEN=1', path_from_root('tests', 'hello_world.c')]
+        command += ['-s', 'BINARYEN_METHOD="' + method + '"']
+        print([method, ' : ', ' '.join(command), ' => ', success])
+        subprocess.check_call(command)
+
+        see_polyfill =  'var WasmJS = ' in open('a.wasm.js').read()
+
+        if method and 'interpret' not in method:
+          assert not see_polyfill, 'verify polyfill was not added - we specified a method, and it does not need it'
+        else:
+          assert see_polyfill, 'we need the polyfill'
+
+        def break_cashew():
+          asm = open('a.wasm.asm.js').read()
+          asm = asm.replace('"almost asm"', '"use asm"; var not_in_asm = [].length + (true || { x: 5 }.x);')
+          asm = asm.replace("'almost asm'", '"use asm"; var not_in_asm = [].length + (true || { x: 5 }.x);')
+          with open('a.wasm.asm.js', 'w') as o: o.write(asm)
+        if method.startswith('interpret-asm2wasm'):
+          try_delete('a.wasm.wast') # we should not need the .wast
+          if not success:
+            break_cashew() # we need cashew
+        elif method.startswith('interpret-s-expr'):
+          try_delete('a.wasm.asm.js') # we should not need the .asm.js
+          if not success:
+            try_delete('a.wasm.wast')
+        elif method.startswith('asmjs'):
+          try_delete('a.wasm.wast') # we should not need the .wast
+          break_cashew() # we don't use cashew, so ok to break it
+          if not success:
+            try_delete('a.wasm.js')
+        elif method.startswith('interpret-binary'):
+          try_delete('a.wasm.wast') # we should not need the .wast
+          try_delete('a.wasm.asm.js') # we should not need the .asm.js
+          if not success:
+            try_delete('a.wasm.wasm')
+        else:
+          1/0
+        proc = subprocess.Popen(NODE_JS + ['a.wasm.js'], stdout=subprocess.PIPE, stderr=subprocess.PIPE)
```

Comment:
Can we start using `shared.execute()` (or some similar common infrastructure) in new code?

Reply:
Generally yes, but here we need the return code so a lower-level API is used.

Reply:
ok, can you use it for the check_call above then?

Reply:
Actually isn't `check_call` good to use?

My understanding is we need a wrapper for the universal newlines problem, but that only applies when we want to read the output - if there's no output, the newlines problem isn't there.

Reply:
Maybe; but if `check_call` is perfect, then we should get rid of `shared.check_call`. Otherwise we should use the latter everywhere; we have a lot of inconsistency right now. At least the logging looks nice. And it looks like `execute` has some special handling for Windows, do we want that here too?

Reply:
Yeah, we should improve this. I *think* we should use the `shared.py` versions in the tools themselves, as they log out better errors (which matters less in the tests). But maybe that's not true. And I'm not sure why we have that weird code in `execute`, also:
````
    cmd[0] = Building.remove_quotes(cmd[0])
````
How about if I open an issue to discuss this?

Reply:
Sure, sounds good. You can go ahead and land this, there will be general cleanup either way.